include/algo/kdtree.h

```diff
@@ -1,20 +1,174 @@
-/**
- * @author Ethan Uppal
- * @copyright Copyright (C) 2024 Ethan Uppal.
- * SPDX-License-Identifier: MIT
- */
-
 #pragma once
+#include <vector>
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <stdexcept>
+#include <limits>
+#include <exception>
+#include <functional>
+
+namespace icp {
+    /**
+     * @brief A k-d tree for efficient nearest neighbor search in k-dimensional space.
+     *
+     */
+    template<typename PointT>
+    class KdTree {
+    public:
+        /**
+         * @brief Construct an empty k-d tree.
+         */
+        KdTree(): root_(nullptr), dim_(PointT::DIM) {}
+
+        /**
+         * @brief Construct and build the k-d tree from a point cloud.
+         * @param points Vector of input points.
+         * @param dim Dimensionality of the space. If -1, defaults to 3.
+         */
+        KdTree(const std::vector<PointT>& points, int dim = -1): root_(nullptr) {
+            build(points, dim);
+        }
+
+        /**
+         * @brief Destructor.
+         */
+        ~KdTree() {
+            clear();
+        }
+
+        /**
+         * @brief Build the k-d tree from a given point set.
+         * @param points Vector of input points.
+         * @param dim Dimensionality of the space. If -1, defaults to 3.
+         */
+        void build(const std::vector<PointT>& points, int dim = -1) {
+            clear();
+            points_ = points;
+
+            dim_ = (dim > 0) ? dim : 3;
+
+            std::vector<int> indices(points.size());
+            std::iota(std::begin(indices), std::end(indices), 0);
+
+            root_ = buildRecursive(indices.data(), static_cast<int>(points.size()), 0);
+        }
+
+        /**
+         * @brief Clear the tree.
+         */
+        void clear() {
+            clearRecursive(root_);
+            root_ = nullptr;
+            points_.clear();
+        }
+
+        /**
+         * @brief Internal node structure of the k-d tree.
+         */
+        struct Node {
+            int idx;        // index to the original point
+            Node* next[2];  // pointers to the child nodes
```

Comment:
If we use smart pointers like `unique_ptr` here we will get memory management for free. I can help with this!